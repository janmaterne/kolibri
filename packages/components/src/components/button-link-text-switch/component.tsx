/* eslint-disable jsx-a11y/label-has-associated-control */
import { Component, h, Host, JSX, Prop } from '@stencil/core';

import { ButtonOrLinkOrTextWithChildrenProps, ButtonWithChildrenProps, LinkWithChildrenProps, TextWithChildrenProps } from '../../types/button-link-text';
import { Props } from './types';
import { KolButtonWcTag, KolLinkWcTag, KolSpanWcTag } from '../../core/component-names';

/**
 * Internal component that renders an action or text component like a button or a link.
 * @internal
 */
@Component({
	tag: 'kol-button-link-text-switch-wc',
	shadow: false,
})
export class KolButtonLinkTextSwitchWc implements Props {
	public render(): JSX.Element {
		return <Host class="kol-button-link-text-switch-wc">{this.renderContent()}</Host>;
	}

	private renderContent() {
		if ((this._link as ButtonWithChildrenProps)._on) {
			const button = this._link as ButtonWithChildrenProps;
			return this.button(button);
		} else if ((this._link as LinkWithChildrenProps)._href) {
			const link = this._link as LinkWithChildrenProps;
			return this.link(link);
		} else {
			const text = this._link as TextWithChildrenProps;
			return this.text(text);
		}
	}

	private button = (button: ButtonWithChildrenProps): JSX.Element => <KolButtonWcTag {...button}></KolButtonWcTag>;

	private link = (link: LinkWithChildrenProps): JSX.Element => <KolLinkWcTag {...link}></KolLinkWcTag>;

	private text = (text: TextWithChildrenProps): JSX.Element => <KolSpanWcTag {...text}></KolSpanWcTag>;

	/**
	 * Die Link-Daten welche diese Komponente verwendet, um die entsprechende Komponente zu rendern.
	 */
	@Prop() public _link!: ButtonOrLinkOrTextWithChildrenProps;
}
