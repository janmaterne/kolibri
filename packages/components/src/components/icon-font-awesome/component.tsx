import type { Generic } from 'adopted-style-sheets';
import { Component, h, Host, JSX, Prop } from '@stencil/core';

import { FontAwesome, FontAwesomeOssPrefix } from '../../enums/font-awesome';
import { PropAriaLabel } from '../../types/props/aria-label';
import { KolIconTag } from '../../core/component-names';

type RequiredProps = PropAriaLabel & {
	icon: FontAwesome;
	prefix: FontAwesomeOssPrefix;
};
type OptionalProps = {
	part: string;
};
// type Props = Generic.Element.Members<RequiredProps, OptionalProps>;

/**
 * @deprecated Wir empfehlen die flexiblere KolIcon-Komponente zu verwenden.
 */
@Component({
	tag: 'kol-icon-font-awesome',
	shadow: false,
})
export class KolIconFontAwesome implements Generic.Element.Members<RequiredProps, OptionalProps> {
	public render(): JSX.Element {
		return (
			<Host class="kol-icon-font-awesome">
				<KolIconTag
					exportparts={`icon${typeof this._part === 'string' ? `,${this._part}` : ''}`}
					_ariaLabel={this._ariaLabel}
					_icon={typeof this._prefix === 'string' && typeof this._icon === 'string' ? `${this._prefix} fa-${this._icon}` : (undefined as unknown as string)}
				/>
			</Host>
		);
	}

	/**
	 * Setzt die semantische Beschriftung der Komponente.
	 */
	@Prop() public _ariaLabel!: string;

	/**
	 * Gibt einen Identifier eines Font Awesome Icons an. (https://fontawesome.com/v5.15/icons)
	 */
	@Prop() public _icon!: FontAwesome;

	/**
	 * Gibt den Prefix der Font Awesome Icons an. (https://fontawesome.com/v5.15/how-to-use/on-the-web/referencing-icons/basic-use)
	 */
	@Prop() public _prefix!: FontAwesomeOssPrefix;

	/**
	 * Deprecated: Gibt den Identifier für den CSS-Part an, um das Icon von Außen ändern zu können. (https://meowni.ca/posts/part-theme-explainer/)
	 *
	 * @deprecated Das Styling sollte stets über CSS erfolgen.
	 */
	@Prop() public _part?: string;
}
