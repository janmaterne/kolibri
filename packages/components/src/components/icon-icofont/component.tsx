import type { Generic } from 'adopted-style-sheets';
import { Component, h, Host, JSX, Prop } from '@stencil/core';

import { Icofont } from '../../types/icofont';
import { PropAriaLabel } from '../../types/props/aria-label';
import { KolIconTag } from '../../core/component-names';

type RequiredProps = PropAriaLabel & {
	icon: Icofont;
};
type OptionalProps = {
	part: string;
};
// type Props = Generic.Element.Members<RequiredProps, OptionalProps>;

/**
 * @deprecated Wir empfehlen die flexiblere KolIcon-Komponente zu verwenden.
 */
@Component({
	tag: 'kol-icon-icofont',
	shadow: false,
})
export class KolIconIcofont implements Generic.Element.Members<RequiredProps, OptionalProps> {
	public render(): JSX.Element {
		return (
			<Host class="kol-icon-icofont">
				<KolIconTag
					exportparts={`icon${typeof this._part === 'string' ? `,${this._part}` : ''}`}
					_ariaLabel={this._ariaLabel}
					_icon={typeof this._icon === 'string' ? `icofont-${this._icon}` : (undefined as unknown as string)}
				/>
			</Host>
		);
	}

	/**
	 * Setzt die semantische Beschriftung der Komponente.
	 */
	@Prop() public _ariaLabel!: string;

	/**
	 * Gibt einen Identifier eines Icofont Icons an. (https://icofont.com/icons)
	 */
	@Prop() public _icon!: Icofont;

	/**
	 * Deprecated: Gibt den Identifier für den CSS-Part an, um das Icon von Außen ändern zu können. (https://meowni.ca/posts/part-theme-explainer/)
	 *
	 * @deprecated Das Styling sollte stets über CSS erfolgen.
	 */
	@Prop() public _part?: string;
}
