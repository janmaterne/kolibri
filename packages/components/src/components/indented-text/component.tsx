import type { IndentedTextAPI, IndentedTextStates } from '../../schema';
import type { JSX } from '@stencil/core';
import { Component, h, Host, State } from '@stencil/core';

/**
 * @slot - Der Text.
 */
@Component({
	tag: 'kol-indented-text-wc',
	shadow: false,
})
export class KolIndentedTextWc implements IndentedTextAPI {
	@State() public state: IndentedTextStates = {};

	public render(): JSX.Element {
		return (
			<Host class="kol-indented-text-wc">
				<div>
					{/* TODO: Für was wird dieses div benötigt? */}
					<slot />
				</div>
			</Host>
		);
	}
}
