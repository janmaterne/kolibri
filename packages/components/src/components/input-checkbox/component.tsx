import { Component, Element, h, Host, JSX, Prop, State, Watch } from '@stencil/core';

import { Stringified } from '../../types/common';
import { InputTypeOnDefault } from '../../types/input/types';
import { LabelWithExpertSlotPropType } from '../../types/props/label';
import { StencilUnknown } from '../../types/unknown';
import { nonce } from '../../utils/dev.utils';
import { stopPropagation, tryToDispatchKoliBriEvent } from '../../utils/events';
import { propagateFocus } from '../../utils/reuse';
import { getRenderStates } from '../input/controller';
import { InputCheckboxController } from './controller';
import { API, InputCheckboxIcon, InputCheckboxVariant, States } from './types';
import { CheckedPropType } from '../../types/props/checked';
import { IndeterminatePropType } from '../../types/props/indeterminate';
import { SyncValueBySelectorPropType } from '../../types/props/sync-value-by-selector';
import { IdPropType } from '../../types/props/id';
import { NamePropType } from '../../types/props/name';
import { TooltipAlignPropType } from '../../types/props/tooltip-align';

/**
 * @slot - Die Beschriftung der Checkbox.
 */
@Component({
	tag: 'kol-input-checkbox',
	styleUrls: {
		default: './style.css',
	},
	shadow: true,
})
export class KolInputCheckbox implements API {
	@Element() private readonly host?: HTMLKolInputCheckboxElement;
	private ref?: HTMLInputElement;

	private readonly catchRef = (ref?: HTMLInputElement) => {
		this.ref = ref;
		propagateFocus(this.host, this.ref);
	};

	public render(): JSX.Element {
		const { ariaDescribedBy } = getRenderStates(this.state);
		const hasExpertSlot = this.state._label === false; // _label="" or _label

		return (
			<Host>
				<kol-input
					class={{
						checkbox: true,
						[this.state._variant]: true,
						'hide-label': !!this.state._hideLabel,
						checked: this.state._checked,
					}}
					data-role={this.state._variant === 'button' ? 'button' : undefined}
					onKeyPress={this.state._variant === 'button' ? this.onChange : undefined}
					tabIndex={this.state._variant === 'button' ? 0 : undefined}
					_alert={this.state._alert}
					_disabled={this.state._disabled}
					_error={this.state._error}
					_hideLabel={this.state._hideLabel}
					_hint={this.state._hint}
					_id={this.state._id}
					_required={this.state._required}
					_touched={this.state._touched}
				>
					{/*  TODO: der folgende Slot ohne Name muss später entfernt werden */}
					<span slot="label">{hasExpertSlot ? <slot></slot> : this.state._label}</span>
					<div slot="input">
						<kol-icon
							onClick={this.onChange}
							_ariaLabel=""
							_icon={this.state._indeterminate ? this.state._icon.indeterminate : this.state._checked ? this.state._icon.checked : this.state._icon.unchecked}
						/>
						<input
							ref={this.catchRef}
							title=""
							accessKey={this.state._accessKey} // by checkbox?!
							aria-describedby={ariaDescribedBy.length > 0 ? ariaDescribedBy.join(' ') : undefined}
							aria-label={this.state._hideLabel && typeof this.state._label === 'string' ? this.state._label : undefined}
							checked={this.state._checked}
							disabled={this.state._disabled}
							id={this.state._id}
							indeterminate={this.state._indeterminate}
							name={this.state._name}
							required={this.state._required}
							tabIndex={this.state._tabIndex}
							type="checkbox"
							{...this.controller.onFacade}
							onChange={this.onChange}
							onClick={undefined} // onClick is not needed since onChange already triggers the correct event
						/>
						<kol-tooltip-wc
							/**
							 * Dieses Aria-Hidden verhindert das doppelte Vorlesen des Labels,
							 * verhindert aber nicht das Aria-Labelledby vorgelesen wird.
							 */
							aria-hidden="true"
							class="input-tooltip"
							hidden={hasExpertSlot || !this.state._hideLabel}
							_align={this._tooltipAlign}
							_label={typeof this.state._label === 'string' ? this.state._label : ''}
						></kol-tooltip-wc>
					</div>
				</kol-input>
			</Host>
		);
	}

	private readonly controller: InputCheckboxController;

	/**
	 * Defines which key combination can be used to trigger or focus the interactive element of the component.
	 */
	@Prop() public _accessKey?: string;

	/**
	 * Defines whether the screen-readers should read out the notification.
	 */
	@Prop({ mutable: true, reflect: true }) public _alert?: boolean = true;

	/**
	 * Defines whether the checkbox is checked or not. Can be read and written.
	 * @TODO: Change type back to `CheckedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _checked?: boolean = false;

	/**
	 * Makes the element not focusable and ignore all events.
	 * @TODO: Change type back to `DisabledPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _disabled?: boolean;

	/**
	 * Defines the error message text.
	 */
	@Prop() public _error?: string;

	/**
	 * Hides the label and shows the description in a Tooltip instead.
	 * @TODO: Change type back to `HideLabelPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hideLabel?: boolean;

	/**
	 * Defines the hint text.
	 */
	@Prop() public _hint?: string = '';

	/**
	 * Defines the icon classnames (e.g. `_icon="fa-solid fa-user"`).
	 */
	@Prop() public _icon?: Stringified<InputCheckboxIcon>;

	/**
	 * Defines the internal ID of the primary component element.
	 */
	@Prop() public _id?: IdPropType;

	/**
	 * Puts the checkbox in the indeterminate state, does not change the value of _checked.
	 * @TODO: Change type back to `IndeterminatePropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _indeterminate?: boolean;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.). Set to `false` to enable the expert slot.
	 */
	@Prop() public _label?: LabelWithExpertSlotPropType;

	/**
	 * Defines the technical name of an input field.
	 */
	@Prop() public _name?: NamePropType;

	/**
	 * Gibt die EventCallback-Funktionen für das Input-Event an.
	 */
	@Prop() public _on?: InputTypeOnDefault;

	/**
	 * Makes the input element required.
	 * @TODO: Change type back to `RequiredPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _required?: boolean;

	/**
	 * Selector for synchronizing the value with another input element.
	 * @internal
	 */
	@Prop() public _syncValueBySelector?: SyncValueBySelectorPropType;

	/**
	 * Defines which tab-index the primary element of the component has. (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex)
	 */
	@Prop() public _tabIndex?: number;

	/**
	 * Defines where to show the Tooltip preferably: top, right, bottom or left.
	 */
	@Prop() public _tooltipAlign?: TooltipAlignPropType = 'top';

	/**
	 * Shows if the input was touched by a user.
	 * @TODO: Change type back to `TouchedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _touched?: boolean = false;

	/**
	 * Deprecated: Defines which variant should be used for presentation.
	 *
	 * @deprecated Verwende stattdessen das Attribute _variant.
	 */
	@Prop() public _type?: InputCheckboxVariant;

	/**
	 * Defines the value of the input.
	 */
	@Prop() public _value?: Stringified<StencilUnknown> = true;

	/**
	 * Defines which variant should be used for presentation.
	 */
	@Prop() public _variant?: InputCheckboxVariant; // TODO: = 'default'; in v2 setzen

	@State() public state: States = {
		_checked: false,
		_icon: {
			checked: 'codicon codicon-check',
			indeterminate: 'codicon codicon-remove',
			unchecked: 'codicon codicon-add',
		},
		_id: `id-${nonce()}`,
		_indeterminate: false,
		_label: false, // ⚠ required
		_value: true,
		_variant: 'default',
	};

	public constructor() {
		this.controller = new InputCheckboxController(this, 'input-checkbox', this.host);
	}

	@Watch('_accessKey')
	public validateAccessKey(value?: string): void {
		this.controller.validateAccessKey(value);
	}

	@Watch('_alert')
	public validateAlert(value?: boolean): void {
		this.controller.validateAlert(value);
	}

	@Watch('_checked')
	public validateChecked(value?: CheckedPropType): void {
		this.controller.validateChecked(value);
	}

	@Watch('_disabled')
	public validateDisabled(value?: boolean): void {
		this.controller.validateDisabled(value);
	}

	@Watch('_error')
	public validateError(value?: string): void {
		this.controller.validateError(value);
	}

	@Watch('_hideLabel')
	public validateHideLabel(value?: boolean): void {
		this.controller.validateHideLabel(value);
	}

	@Watch('_hint')
	public validateHint(value?: string): void {
		this.controller.validateHint(value);
	}

	@Watch('_icon')
	public validateIcon(value?: Stringified<InputCheckboxIcon>): void {
		this.controller.validateIcon(value);
	}

	@Watch('_id')
	public validateId(value?: string): void {
		this.controller.validateId(value);
	}

	@Watch('_indeterminate')
	public validateIndeterminate(value?: IndeterminatePropType): void {
		this.controller.validateIndeterminate(value);
	}

	@Watch('_label')
	public validateLabel(value?: LabelWithExpertSlotPropType): void {
		this.controller.validateLabel(value);
	}

	@Watch('_name')
	public validateName(value?: string): void {
		this.controller.validateName(value);
	}

	@Watch('_on')
	public validateOn(value?: InputTypeOnDefault): void {
		this.controller.validateOn(value);
	}

	@Watch('_required')
	public validateRequired(value?: boolean): void {
		this.controller.validateRequired(value);
	}

	@Watch('_syncValueBySelector')
	public validateSyncValueBySelector(value?: SyncValueBySelectorPropType): void {
		this.controller.validateSyncValueBySelector(value);
	}

	@Watch('_tabIndex')
	public validateTabIndex(value?: number): void {
		this.controller.validateTabIndex(value);
	}

	@Watch('_touched')
	public validateTouched(value?: boolean): void {
		this.controller.validateTouched(value);
	}

	@Watch('_type')
	public validateType(value?: InputCheckboxVariant): void {
		this.controller.validateType(value);
	}

	@Watch('_value')
	public validateValue(value?: Stringified<StencilUnknown>): void {
		this.controller.validateValue(value);
	}

	@Watch('_variant')
	public validateVariant(value?: InputCheckboxVariant): void {
		this.controller.validateVariant(value);
	}

	public componentWillLoad(): void {
		this._alert = this._alert === true;
		this._touched = this._touched === true;
		this.controller.componentWillLoad();
	}

	private onChange = (event: Event): void => {
		this._checked = !this._checked;
		this._indeterminate = false;

		const value = this._checked ? this.state._value : null;

		// Event handling
		stopPropagation(event);
		tryToDispatchKoliBriEvent('change', this.host, value);

		// Static form handling
		this.controller.setFormAssociatedValue(value);
		this.controller.setFormAssociatedCheckboxValue;

		// Callback
		if (typeof this.state._on?.onChange === 'function') {
			this.state._on.onChange(event, value);
		}
	};
}
