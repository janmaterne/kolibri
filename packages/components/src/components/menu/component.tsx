/* eslint-disable @typescript-eslint/no-explicit-any */
import { h, Component, Host, Prop, State, Watch, type JSX } from '@stencil/core';
import type { MenuAPI, MenuItemProps, MenuStates, Stringified } from '../../schema';

enum KeyMap {
	ArrowDown = 'ArrowDown',
	ArrowUp = 'ArrowUp',
	Enter = 'Enter',
	Spacebar = '',
}

@Component({
	tag: 'kol-menu-wc',
	styleUrl: './style.scss',
	shadow: false,
})
export class KolMenu implements MenuAPI {
	private readonly renderMenuItem = (item: MenuItemProps, index: number): JSX.Element => {
		// ref={(el) => (this.listItems[index] = el!)}
		// tabindex={this.state._pageSize === option.value ? '0' : '-1'}
		// key={option.value}

		return (
			<li role="menuitem" onClick={(event) => this.fireOnChange(event, item)} onKeyDown={(event) => this.handleKeyDown(event, index, item)}>
				{item._label}
			</li>
		);
	};

	private handleKeyDown(event: KeyboardEvent, index: number, item: unknown): void {
		console.log('INDEX: ', index);
		switch (event.key as KeyMap) {
			case KeyMap.ArrowDown:
				this.selectNextMenuItem();
				event.preventDefault();
				break;
			case KeyMap.ArrowUp:
				this.selectPrevMenuItem();
				event.preventDefault();
				break;
			case KeyMap.Enter:
			case KeyMap.Spacebar:
				this.fireOnChange(event, item);
				event.preventDefault();
				break;
			default:
				break;
		}
	}

	private selectNextMenuItem() {
		// const nextIndex = index + 1 < this.listItems.length ? index + 1 : 0;
		//	this.listItems[nextIndex].focus();
	}

	private selectPrevMenuItem() {
		// const prevIndex = index - 1 >= 0 ? index - 1 : this.listItems.length - 1;
		// this.listItems[prevIndex].focus();
	}

	private fireOnChange(event: unknown, item: unknown) {
		console.log('fireOnChange', { event, item });
		// this.onChangePageSize(event, value);
	}

	public render(): JSX.Element {
		console.log('RENDER: ', this._menuItems, this.state._menuItems);

		return (
			<Host class="kol-menu-wc">
				<ul class="menu">{(this._menuItems as any[])?.map(this.renderMenuItem) || <div></div>}</ul>
			</Host>
		);
	}

	/**
	 *
	 */
	@Prop() public _menuItems: Stringified<MenuItemProps[]> = [];

	@State() public state: MenuStates = {
		_menuItems: [],
	};

	@Watch('_menuItems')
	public validateMenuItems(value?: Stringified<MenuItemProps[]>): void {
		console.log('value: ', value);
		// watchNavLinks('KolBreadcrumb', this, value);
	}

	public componentWillLoad(): void {
		console.log('componentWillLoad');
		this.validateMenuItems(this._menuItems);
	}
}

/**
 <ul class="dropdown-menu">
							{this.state._pageSizeOptions.map((option, index) => (
								<li
									ref={(el) => (this.listItems[index] = el!)}
									key={option.value}
									role="menuitem"
									tabindex={this.state._pageSize === option.value ? '0' : '-1'}
									aria-selected={this.state._pageSize === option.value ? 'true' : 'false'}
									onClick={(event) => {
										this.onChangePageSize(event, option.value);
									}}
									onKeyDown={(event) => this.handleKeyDown(event, index, option.value)}
								>
									{option.label} {this.state._pageSize === option.value && <span> </span>}
								</li>
							))}
						</ul>
 */

/**
private handleKeyDown(event: KeyboardEvent, index: number, value: unknown): void {
		if (event.key === 'ArrowDown') {
			const nextIndex = index + 1 < this.listItems.length ? index + 1 : 0;
			this.listItems[nextIndex].focus();
			event.preventDefault();
		} else if (event.key === 'ArrowUp') {
			const prevIndex = index - 1 >= 0 ? index - 1 : this.listItems.length - 1;
			this.listItems[prevIndex].focus();
			event.preventDefault();
		} else if (event.key === 'Enter' || event.key === ' ') {
			this.onChangePageSize(event, value);
			event.preventDefault();
		}
	} 
*/
