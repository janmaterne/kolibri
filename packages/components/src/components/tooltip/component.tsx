import { arrow, autoUpdate, computePosition, flip, offset, shift } from '@floating-ui/dom';
import { Component, Element, h, Host, JSX, Prop, State, Watch } from '@stencil/core';

import { AlignPropType, validateAlign } from '../../types/props/align';
import { IdPropType, validateId } from '../../types/props/id';
import { LabelPropType, validateLabel } from '../../types/props/label';
import { getDocument } from '../../utils/dev.utils';
import { hideOverlay, showOverlay } from '../../utils/overlay';
import { processEnv } from '../../utils/reuse';
import { API, States } from './types';
import { KolSpanWcTag } from '../../core/component-names';

@Component({
	tag: 'kol-tooltip-wc',
	styleUrl: './style.scss',
	shadow: false,
})
export class KolTooltipWc implements API {
	@Element() private host!: HTMLKolTooltipWcElement;

	private arrowElement?: HTMLDivElement;
	private previousSibling?: Element | null;
	private tooltipElement?: HTMLDivElement;
	private hasFocusIn = false;
	private hasMouseIn = false;

	private cleanupAutoPositioning?: () => void;

	private alignTooltip = (): void => {
		if (processEnv !== 'test' && this.previousSibling /* SSR instanceof HTMLElement */ && this.tooltipElement /* SSR instanceof HTMLElement */) {
			const target = this.previousSibling;
			const tooltipEl = this.tooltipElement;
			const arrowEl = this.arrowElement;

			const middleware = [offset(arrowEl?.offsetHeight ?? 10), flip(), shift()];
			if (arrowEl) {
				middleware.push(arrow({ element: arrowEl }));
			}

			void computePosition(target, tooltipEl, {
				placement: this.state._align,
				middleware: middleware,
			}).then(({ x, y, middlewareData, placement }) => {
				Object.assign(tooltipEl.style, {
					left: `${x}px`,
					top: `${y}px`,
					visibility: 'visible',
				});

				if (arrowEl) {
					if (middlewareData.arrow?.x) {
						Object.assign(arrowEl.style, {
							left: `${middlewareData.arrow.x}px`,
							top: placement === 'bottom' ? `${-arrowEl.offsetHeight / 2}px` : '',
							bottom: placement === 'top' ? `${-arrowEl.offsetHeight / 2}px` : '',
						});
					} else if (middlewareData.arrow?.y) {
						Object.assign(arrowEl.style, {
							left: placement === 'right' ? `${-arrowEl.offsetWidth / 2}px` : '',
							right: placement === 'left' ? `${-arrowEl.offsetWidth / 2}px` : '',
							top: `${middlewareData.arrow.y}px`,
						});
					}
				}
			});
		}
	};

	private showTooltip = (): void => {
		if (this.previousSibling && this.tooltipElement /* SSR instanceof HTMLElement */) {
			showOverlay(this.tooltipElement);
			this.tooltipElement.style.setProperty('display', 'block');
			getDocument().addEventListener('keyup', this.hideTooltipByEscape);

			const target = this.previousSibling;
			const tooltipEl = this.tooltipElement;
			this.cleanupAutoPositioning = autoUpdate(target, tooltipEl, this.alignTooltip);
		}
	};

	private hideTooltip = (): void => {
		if (this.tooltipElement /* SSR instanceof HTMLElement */) {
			hideOverlay(this.tooltipElement);
			this.tooltipElement.style.setProperty('display', 'none');
			this.tooltipElement.style.setProperty('visibility', 'hidden');
			if (this.cleanupAutoPositioning) {
				this.cleanupAutoPositioning();
				this.cleanupAutoPositioning = undefined;
			}
		}
		getDocument().removeEventListener('keyup', this.hideTooltipByEscape);
	};

	private hideTooltipByEscape = (event: KeyboardEvent): void => {
		if (event.key === 'Escape') {
			this.hideTooltip();
		}
	};

	private handleMouseEnter() {
		this.hasMouseIn = true;
		this.showOrHideTooltip();
	}
	private handleMouseleave() {
		this.hasMouseIn = false;
		this.showOrHideTooltip();
	}
	private handleFocusIn() {
		this.hasFocusIn = true;
		this.showOrHideTooltip();
	}
	private handleFocusout() {
		this.hasFocusIn = false;
		this.showOrHideTooltip();
	}

	private addListeners = (el: Element): void => {
		el.addEventListener('mouseenter', this.handleMouseEnter.bind(this));
		el.addEventListener('mouseleave', this.handleMouseleave.bind(this));
		el.addEventListener('focusin', this.handleFocusIn.bind(this));
		el.addEventListener('focusout', this.handleFocusout.bind(this));
	};

	private removeListeners = (el: Element): void => {
		el.removeEventListener('mouseenter', this.handleMouseEnter.bind(this));
		el.removeEventListener('mouseleave', this.handleMouseleave.bind(this));
		el.removeEventListener('focusin', this.handleFocusIn.bind(this));
		el.removeEventListener('focusout', this.handleFocusout.bind(this));
	};

	private resyncListeners = (last?: Element | null, next?: Element | null, replacePreviousSibling = false): void => {
		if (last) {
			this.removeListeners(last);
		}
		if (next) {
			/**
			 * This makes the next element to the last element for the next resync cycle.
			 */
			if (replacePreviousSibling) {
				this.previousSibling = next;
			}
			this.addListeners(next);
		}
	};

	private catchTooltipElement = (el?: HTMLDivElement): void => {
		this.tooltipElement = el;
	};
	private catchArrowElement = (element?: HTMLDivElement): void => {
		this.arrowElement = element;
	};

	public render(): JSX.Element {
		return (
			<Host class="kol-tooltip-wc">
				{this.state._label !== '' && (
					<div class="tooltip-floating" ref={this.catchTooltipElement}>
						<div class="tooltip-area tooltip-arrow" ref={this.catchArrowElement} />
						<KolSpanWcTag class="tooltip-area tooltip-content" id={this.state._id} _label={this.state._label}></KolSpanWcTag>
					</div>
				)}
			</Host>
		);
	}

	/**
	 * Defines the alignment of the tooltip, popover or tabs in relation to the element.
	 */
	@Prop() public _align?: AlignPropType = 'top';

	/**
	 * Defines the internal ID of the primary component element.
	 */
	@Prop() public _id?: IdPropType;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.).
	 */
	@Prop() public _label!: LabelPropType;

	@State() public state: States = {
		_align: 'top',
		_label: '…', // ⚠ required
	};

	@Watch('_align')
	public validateAlign(value?: AlignPropType): void {
		validateAlign(this, value);
	}

	@Watch('_id')
	public validateId(value?: IdPropType): void {
		validateId(this, value);
	}

	@Watch('_label')
	public validateLabel(value?: LabelPropType): void {
		validateLabel(this, value, {
			required: true,
		});
	}

	private overFocusTimeout?: ReturnType<typeof setTimeout>;

	private showOrHideTooltip = (): void => {
		clearTimeout(this.overFocusTimeout);
		this.overFocusTimeout = setTimeout(() => {
			if (this.hasMouseIn || this.hasFocusIn) {
				this.showTooltip();
			} else {
				this.hideTooltip();
			}
		}, 250);
	};

	public componentWillLoad(): void {
		this.validateAlign(this._align);
		this.validateId(this._id);
		this.validateLabel(this._label);
	}

	private handleEventListeners(): void {
		this.resyncListeners(this.previousSibling, this.host?.previousElementSibling, true);
		this.resyncListeners(this.tooltipElement, this.tooltipElement);
	}

	public connectedCallback(): void {
		this.previousSibling = this.host?.previousElementSibling;
	}

	public componentDidRender(): void {
		this.handleEventListeners();
	}

	/**
	 * @see: components/abbr/component.tsx (componentDidLoad)
	 */
	public disconnectedCallback(): void {
		if (this.previousSibling /* SSR instanceof HTMLElement */) {
			this.removeListeners(this.previousSibling);
			this.previousSibling = undefined;
		}
		if (this.tooltipElement /* SSR instanceof HTMLElement */) {
			this.removeListeners(this.tooltipElement);
		}
		if (this.cleanupAutoPositioning) {
			this.cleanupAutoPositioning();
		}
	}
}
