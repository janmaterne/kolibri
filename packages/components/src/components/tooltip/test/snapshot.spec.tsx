import { KolTooltipWcTag } from '../../../core/component-names';
import type { TooltipProps } from '../../../schema';
import { executeSnapshotTests } from '../../../utils/testing';

import { KolTooltipWc } from '../component';

executeSnapshotTests<TooltipProps>(
	KolTooltipWcTag,
	[KolTooltipWc],
	[
		{ _id: 'id', _label: 'Label' },
		{ _id: 'id', _label: 'Label', _align: 'top' },
		{ _id: 'id', _label: 'Label', _align: 'left' },
		{ _id: 'id', _label: 'Label', _align: 'right' },
		{ _id: 'id', _label: 'Label', _align: 'bottom' },
	],
);
