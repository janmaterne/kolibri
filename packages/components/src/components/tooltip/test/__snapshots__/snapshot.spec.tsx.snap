// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kol-tooltip-wc should render with _id="id" _label="Label" _align="bottom" 1`] = `
<kol-tooltip-wc class="kol-tooltip-wc">
  <div class="tooltip-floating">
    <div class="tooltip-area tooltip-arrow"></div>
    <kol-span-wc _label="Label" class="tooltip-area tooltip-content" id="id"></kol-span-wc>
  </div>
</kol-tooltip-wc>
`;

exports[`kol-tooltip-wc should render with _id="id" _label="Label" _align="left" 1`] = `
<kol-tooltip-wc class="kol-tooltip-wc">
  <div class="tooltip-floating">
    <div class="tooltip-area tooltip-arrow"></div>
    <kol-span-wc _label="Label" class="tooltip-area tooltip-content" id="id"></kol-span-wc>
  </div>
</kol-tooltip-wc>
`;

exports[`kol-tooltip-wc should render with _id="id" _label="Label" _align="right" 1`] = `
<kol-tooltip-wc class="kol-tooltip-wc">
  <div class="tooltip-floating">
    <div class="tooltip-area tooltip-arrow"></div>
    <kol-span-wc _label="Label" class="tooltip-area tooltip-content" id="id"></kol-span-wc>
  </div>
</kol-tooltip-wc>
`;

exports[`kol-tooltip-wc should render with _id="id" _label="Label" _align="top" 1`] = `
<kol-tooltip-wc class="kol-tooltip-wc">
  <div class="tooltip-floating">
    <div class="tooltip-area tooltip-arrow"></div>
    <kol-span-wc _label="Label" class="tooltip-area tooltip-content" id="id"></kol-span-wc>
  </div>
</kol-tooltip-wc>
`;

exports[`kol-tooltip-wc should render with _id="id" _label="Label" 1`] = `
<kol-tooltip-wc class="kol-tooltip-wc">
  <div class="tooltip-floating">
    <div class="tooltip-area tooltip-arrow"></div>
    <kol-span-wc _label="Label" class="tooltip-area tooltip-content" id="id"></kol-span-wc>
  </div>
</kol-tooltip-wc>
`;
